### ISSUE TITLE
Issue: YOLO classification model  Forces Image Resize to 640×640 Instead of 320×640 During Training

### ISSUE THREAD

**Issue by AfaqKarim** [ultralytics/ultralytics#22851]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

About a year ago, I trained a YOLOv8 classification model using images with a resolution of 320×640, and I used the parameter rect=True.

Now, when I try to train a new model using the same approach, YOLO automatically resizes the images to 640×640, and I can’t get it to keep the original 320×640 size.

Could anyone guide me on how to fix this and train the model using 320×640 dimensions?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @AfaqKarim, thank you for opening this issue and for using Ultralytics 🚀! This is an **automated initial response** to help you gather the right information quickly. An Ultralytics engineer will review your report and assist you in more detail soon 😊

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, including configuration options that affect image sizes and dataloading behavior. Many common training and classification questions are already answered there.

Since your report involves training behavior that changed over time and may indicate a regression or configuration issue:

- If this is a 🐛 **Bug Report**, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE), including:
  - Your exact training command (CLI) or Python script
  - The `model`, `imgsz`, and `rect` settings you are using
  - A small sample of your dataset structure (paths and labels; a tiny subset of images if possible)
  - Full console logs from the start of training up to the point where you observe the 640×640 resizing
  - Your environment details (OS, Python version, PyTorch version, CUDA, GPU)

This will help us confirm whether the forced 640×640 resizing is a bug or an expected change in current behavior.

- If this is a custom training ❓ **Question**, please provide as much information as possible, including:
  - A description of your classification dataset (number of classes, example images)
  - The training command or script you previously used (that worked with 320×640)
  - The new command or script you are using now
  - Any screenshots or logs that show the reported 640×640 resizing
  - Confirmation that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)

Join the Ultralytics community where it suits you best for additional help and discussion:
- For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧  
- For in-depth, forum-style discussions, visit [Discourse](https://community.ultralytics.com/)  
- To share your experience and learn from others, check out our [Subreddit](https://www.reddit.com/r/Ultralytics/)  

## Upgrade

Before we dive deeper, please upgrade to the latest `ultralytics` package (which includes YOLO11 and updated classification behavior) with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) installed in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). This helps ensure the behavior you’re seeing is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

After upgrading, try your training again and let us know:
- The exact command used
- Whether the image size is still forced to 640×640
- Any warnings or notes printed at startup

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

If possible, please try to reproduce the behavior in one of these environments and share:
- Which environment you used
- The exact steps taken there
- Whether the classification model still resizes to 640×640

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Once you share the requested details and (ideally) a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/), an Ultralytics engineer will be able to investigate the 320×640 vs 640×640 behavior more precisely and follow up with you 👀

**Comment by Y-T-G** [comment]
`rect` argument is not applicable or supported for classification model training. It applies to other non-classification tasks. `rect` was never supported for classification model training. You probably trained an object detection model before, not classification.

**Comment by Y-T-G** [comment]
If you want to avoid cropped images during training, you can use the code provided here for training:
https://docs.ultralytics.com/tasks/classify/#train

**Comment by AfaqKarim** [comment]
> `rect` argument is not applicable or supported for classification model training. It applies to other non-classification tasks. `rect` was never supported for classification model training. You probably trained an object detection model before, not classification.

Last year, rect was supported—I trained two classification models with different datasets using it.

**Comment by Y-T-G** [comment]
I am fairly confident it was not supported for classification training last year. I think you were using an older version than that. If you have the `.pt` files for the old models, you can load and check Ultralytics version used to train them and then install that version:

```python
from ultralytics import YOLO
model = YOLO("best.pt")  # load your old model
print(model.ckpt["version"])
```

**Comment by Y-T-G** [comment]
It used to be supported for 8.0.190 which is over 2 years old, not from last year.

https://github.com/ultralytics/ultralytics/pull/5162

**Comment by AfaqKarim** [comment]
python test.py 
/home/nexaeye-isb/miniconda3/envs/cip/lib/python3.10/site-packages/ultralytics/nn/tasks.py:578: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  return torch.load(file, map_location='cpu'), file  # load
8.0.207

**Comment by Y-T-G** [comment]
It's an old version that's not supported anymore

You can downgrade PyTorch to 1.12

**Comment by AfaqKarim** [comment]
python test.py 
dict_keys(['epoch', 'best_fitness', 'model', 'ema', 'updates', 'optimizer', 'train_args', 'train_metrics', 'train_results', 'date', 'version'])
Ultralytics version: 8.0.230
Training args: {'task': 'classify', 'mode': 'train', 'model': 'yolov8l-cls.pt', 'data': 'dataset_25feb', 'epochs': 10, 'time': None, 'patience': 50, 'batch': 16, 'imgsz': 640, 'save': True, 'save_period': -1, 'cache': False, 'device': None, 'workers': 8, 'project': 'demo_MODEL', 'name': 'train', 'exist_ok': False, 'pretrained': True, 'optimizer': 'auto', 'verbose': True, 'seed': 0, 'deterministic': True, 'single_cls': False, 'rect': True, 'cos_lr': False, 'close_mosaic': 10, 'resume': False, 'amp': True, 'fraction': 1.0, 'profile': False, 'freeze': None, 'overlap_mask': True, 'mask_ratio': 4, 'dropout': 0.0, 'val': True, 'split': 'val', 'save_json': False, 'save_hybrid': False, 'conf': None, 'iou': 0.7, 'max_det': 300, 'half': False, 'dnn': False, 'plots': True, 'source': None, 'vid_stride': 1, 'stream_buffer': False, 'visualize': False, 'augment': False, 'agnostic_nms': False, 'classes': None, 'retina_masks': False, 'embed': None, 'show': False, 'save_frames': False, 'save_txt': False, 'save_conf': False, 'save_crop': False, 'show_labels': True, 'show_conf': True, 'show_boxes': True, 'line_width': None, 'format': 'torchscript', 'keras': False, 'optimize': False, 'int8': False, 'dynamic': False, 'simplify': False, 'opset': None, 'workspace': 4, 'nms': False, 'lr0': 0.01, 'lrf': 0.01, 'momentum': 0.937, 'weight_decay': 0.0005, 'warmup_epochs': 3.0, 'warmup_momentum': 0.8, 'warmup_bias_lr': 0.0, 'box': 7.5, 'cls': 0.5, 'dfl': 1.5, 'pose': 12.0, 'kobj': 1.0, 'label_smoothing': 0.0, 'nbs': 64, 'hsv_h': 0.015, 'hsv_s': 0.7, 'hsv_v': 0.4, 'degrees': 0.0, 'translate': 0.1, 'scale': 0.5, 'shear': 0.0, 'perspective': 0.0, 'flipud': 0.0, 'fliplr': 0.5, 'mosaic': 1.0, 'mixup': 0.0, 'copy_paste': 0.0, 'cfg': None, 'tracker': 'botsort.yaml'}


this is another model in which i used rect=true

**Comment by glenn-jocher** [comment]
Your `8.0.230` checkpoint shows it was trained with `imgsz=640`, so even there `rect=True` wasn’t actually giving a fixed `320×640` tensor for classification; in current Ultralytics YOLO classification, the pipeline is square by default and `rect` is ignored, so you’ll only get `imgsz×imgsz` unless you override the classification transforms.

If you need true `320×640` inputs (no crop), you can do it in Python by swapping the dataset transforms to a letterbox resize (see the [`ClassifyLetterBox` reference](https://docs.ultralytics.com/reference/data/augment/#classifyletterbox)) and training with a custom `ClassificationTrainer`:

```python
import numpy as np
from PIL import Image
import torchvision.transforms as T

from ultralytics import YOLO
from ultralytics.data.augment import ClassifyLetterBox, DEFAULT_MEAN, DEFAULT_STD
from ultralytics.data.dataset import ClassificationDataset
from ultralytics.models.yolo.classify import ClassificationTrainer, ClassificationValidator


class PILLetterBox:
    def __init__(self, size=(320, 640)):
        self.lb = ClassifyLetterBox(size=size)

    def __call__(self, im):
        im = np.asarray(im) if isinstance(im, Image.Image) else im
        return self.lb(im)  # returns HWC uint8 np.ndarray


class RectClsDataset(ClassificationDataset):
    def __init__(self, root, args, augment=False, prefix=""):
        super().__init__(root, args, augment, prefix)
        self.torch_transforms = T.Compose([
            PILLetterBox((320, 640)),
            T.ToTensor(),
            T.Normalize(mean=DEFAULT_MEAN, std=DEFAULT_STD),
        ])


class RectClsTrainer(ClassificationTrainer):
    def build_dataset(self, img_path, mode="train", batch=None):
        return RectClsDataset(root=img_path, args=self.args, augment=(mode == "train"), prefix=mode)


class RectClsValidator(ClassificationValidator):
    def build_dataset(self, img_path, mode="train"):
        return RectClsDataset(root=img_path, args=self.args, augment=(mode == "train"), prefix=self.args.split)


model = YOLO("yolo11n-cls.pt")
model.train(data="your_dataset", trainer=RectClsTrainer, imgsz=640)  # imgsz is ignored by the custom resize above
```

If you still hit unexpected resizing after this, please open a new issue (this one is closed) and paste `yolo checks` plus the full traceback as text so we can reproduce it.
